File: regions/system/library/swift/retry/Sources/K1s0Retry/RetryConfig.swift
import Foundation

/// リトライ動作を制御する設定。
public struct RetryConfig: Sendable {
    /// 最大試行回数。
    public let maxAttempts: Int
    /// 最初の遅延時間（秒）。
    public let initialDelay: TimeInterval
    /// 最大遅延時間（秒）。
    public let maxDelay: TimeInterval
    /// 遅延時間の乗数（指数バックオフ）。
    public let multiplier: Double
    /// ジッターを適用するか。
    public let jitter: Bool

    public init(
        maxAttempts: Int = 3,
        initialDelay: TimeInterval = 0.1,
        maxDelay: TimeInterval = 30.0,
        multiplier: Double = 2.0,
        jitter: Bool = false
    ) {
        self.maxAttempts = maxAttempts
        self.initialDelay = initialDelay
        self.maxDelay = maxDelay
        self.multiplier = multiplier
        self.jitter = jitter
    }

    /// デフォルトのリトライ設定。
    public static let `default` = RetryConfig()
}

File: regions/system/library/csharp/retry/RetryConfig.cs
namespace K1s0.System.Retry;

public record RetryConfig(
    int MaxAttempts = 3,
    TimeSpan? InitialDelay = null,
    TimeSpan? MaxDelay = null,
    double Multiplier = 2.0,
    bool Jitter = true)
{
    private static readonly Random Rng = new();

    public TimeSpan ComputeDelay(int attempt)
    {
        var initial = InitialDelay ?? TimeSpan.FromMilliseconds(100);
        var max = MaxDelay ?? TimeSpan.FromSeconds(30);

        var delayMs = initial.TotalMilliseconds * Math.Pow(Multiplier, attempt);

        if (Jitter)
        {
            delayMs *= 0.5 + Rng.NextDouble();
        }

        delayMs = Math.Min(delayMs, max.TotalMilliseconds);
        return TimeSpan.FromMilliseconds(Math.Max(0, delayMs));
    }
}
